### PR TITLE
Make logger more configurable

### PR DESCRIPTION
**Pull request by lmcgupe**
Signed-off-by: guy perron <guy.perron@ericsson.com>

### CODE REVIEW

File: packages/core/src/common/logger.ts [hunk 1]
```diff
@@ -214,6 +217,7 @@ export class Logger implements ILogger {
         @inject(ILoggerServer) protected readonly server: ILoggerServer,
         @inject(LoggerWatcher) protected readonly loggerWatcher: LoggerWatcher,
         @inject(LoggerFactory) protected readonly factory: LoggerFactory,
+        @inject(LogServerOptions) @optional() serveroptions: object | undefined,
         @inject(LoggerOptions) @optional() options: object | undefined) {
```
Comment (akosyakov) [inline packages/core/src/common/logger.ts]: Why do we need 2 options? Could we just reuse `options`?

Comment (lmcgupe) [inline packages/core/src/common/logger.ts]: one is for the top level server, the other for children

Comment (epatpol) [inline packages/core/src/common/logger.ts]: Maybe it should be renamed to childrenOptions then?

File: packages/core/src/common/logger.ts [hunk 2]
```diff
@@ -214,11 +217,12 @@ export class Logger implements ILogger {
         @inject(ILoggerServer) protected readonly server: ILoggerServer,
         @inject(LoggerWatcher) protected readonly loggerWatcher: LoggerWatcher,
         @inject(LoggerFactory) protected readonly factory: LoggerFactory,
-        @inject(LoggerOptions) @optional() options: object | undefined) {
+        @inject(LogServerOptions) @optional() serveroptions: object | undefined,
```
Comment (epatpol) [inline packages/core/src/common/logger.ts]: Should be serverOptions to be consistent with this._serverOptions

File: packages/core/src/common/logger.ts [hunk 3]
```diff
@@ -214,11 +217,12 @@ export class Logger implements ILogger {
...
+        @inject(LoggerOptions) @optional() childrenoptions: object | undefined) {
```
Comment (epatpol) [inline packages/core/src/common/logger.ts]: Same thing here

File: packages/core/src/node/bunyan-logger-server.ts [hunk 4]
```diff
@@ -42,11 +42,12 @@ export class BunyanLoggerServer implements ILoggerServer {
         if (['trace', 'debug', 'info', 'warn', 'error', 'fatal'].indexOf(logLevel) < 0) {
             logLevel = 'info';
         }
+    }
 
-        this.loggers.push(bunyan.createLogger({
-            name: 'Theia',
-            level: logLevel
-        }));
+    create(serveroptions: object | undefined) {
```
Comment (akosyakov) [inline packages/core/src/node/bunyan-logger-server.ts]: I don't think you need this method, but an option should be injected into the constructor of `BunyanLoggerServer`. This option should be bounded as a constant value in `loggerBackendModule'.

File: packages/core/src/common/logger.ts [hunk 5]
```diff
@@ -232,8 +236,13 @@ export class Logger implements ILogger {
                 }
             });
         });
+       if (serverOptions !== undefined) {
+          this._serverOptions = serverOptions;
+          server.create(serverOptions);
```
Comment (akosyakov) [inline packages/core/src/common/logger.ts]: I am not sure how it is supposed to work from the client-side. I don't think that class should be changed.

**Comment by akosyakov** [comment]
Related to https://github.com/theia-ide/theia/issues/224

File: packages/core/src/node/bunyan-logger-server.ts (hunk 4 above)
Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: The idea was to make this configurable from the frontend like the childs...
I'm not sure if we really need to however...

Comment (akosyakov) [inline packages/core/src/node/bunyan-logger-server.ts]: that's also my wonder, why do we need an extra remote call on the startup if we can statically configure it on the backend? but if we go with an extra call at least all methods in `Logger` should wait till `create` is resolved

Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: Indeed I think it's better to have ot satically in the backend like you suggested

File: packages/core/src/common/logger.ts (hunk 5 above)
Comment (lmcgupe) [inline packages/core/src/common/logger.ts]: removed

File: packages/core/src/node/bunyan-logger-server.ts (hunk 4 above)
Comment (lmcgupe) [inline packages/core/src/node/bunyan-logger-server.ts]: i made the change, it is now statically injected.

File: packages/core/src/common/logger.ts [hunk 6]
```diff
@@ -26,6 +26,7 @@ export type LoggerFactory = (options?: object) => ILogger;
 
 export const LoggerOptions = Symbol('LoggerOptions')
 
+export const LogServerOptions = Symbol('LogServerOptions')
```
Comment (hexa00) [inline packages/core/src/common/logger.ts]: Should be called LoggerServerOptions

File: packages/core/src/node/bunyan-logger-server.ts [hunk 7]
```diff
@@ -32,7 +33,9 @@ export class BunyanLoggerServer implements ILoggerServer {
     /* Root logger id, this is a workaround to a bug with 0 in jsonrpc.  */
     private readonly rootLoggerId = 1;
 
-    constructor() {
+    constructor(
+        @inject(LogServerOptions) serverOptions: object | undefined
```
Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: This is not optional so no | undefined

File: packages/core/src/node/bunyan-logger-server.ts [hunk 8]
```diff
@@ -42,11 +45,9 @@ export class BunyanLoggerServer implements ILoggerServer {
...
             logLevel = 'info';
         }
-
-        this.loggers.push(bunyan.createLogger({
-            name: 'Theia',
-            level: logLevel
-        }));
+        this.loggers.push(bunyan.createLogger(
+            JSON.parse(JSON.stringify(serverOptions))
+        ));
```
Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: So this takes an object, makes it a string and then makes it an object again...? I fail to see the point

File: packages/core/src/node/logger-backend-module.ts [hunk 9]
```diff
@@ -16,12 +16,12 @@ export const loggerBackendModule = new ContainerModule(bind => {
     bind(ILogger).to(Logger).inSingletonScope();
     bind(LoggerWatcher).toSelf().inSingletonScope();
     bind(ILoggerServer).to(BunyanLoggerServer).inSingletonScope();
+    bind(LogServerOptions).toConstantValue(JSON.parse(('{"name": "Theia", "level": "info"}')));
```
Comment (hexa00) [inline packages/core/src/node/logger-backend-module.ts]: No need to parse a string here, just write an object

File: packages/core/src/node/logger-backend-module.ts [hunk 10]
```diff
@@ -16,12 +16,12 @@ export const loggerBackendModule = new ContainerModule(bind => {
...
     bind(LoggerFactory).toFactory(ctx =>
         (options?: any) => {
             const child = new Container({ defaultScope: 'Singleton' });
             child.parent = ctx.container;
             child.bind(ILogger).to(Logger).inTransientScope();
-            child.bind(LoggerOptions).toConstantValue(options);
```
Comment (hexa00) [inline packages/core/src/node/logger-backend-module.ts]: why is this removed?

File: packages/core/src/common/logger.ts (hunk 6 above)
Comment (hexa00) [inline packages/core/src/common/logger.ts]: And it should go in logger-protocol.ts

File: packages/core/src/node/bunyan-logger-server.ts (hunk 7 above)
Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: it also could be on the same line (@inject...) since there's only one

Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: You can just call it options too

File: packages/core/src/node/logger-backend-module.ts [hunk 11]
```diff
@@ -7,7 +7,7 @@
 
 import { ContainerModule, Container } from 'inversify';
 import { ConnectionHandler, JsonRpcConnectionHandler } from "../common/messaging";
-import { ILogger, LoggerFactory, LoggerOptions, Logger } from '../common/logger';
+import { ILogger, LoggerFactory, Logger, LogServerOptions } from '../common/logger';
 import { ILoggerServer, ILoggerClient, loggerPath } from '../common/logger-protocol';
```
Comment (hexa00) [inline packages/core/src/node/logger-backend-module.ts]: LoggerOptions needs to stay

File: packages/core/src/node/bunyan-logger-server.ts [hunk 12]
```diff
@@ -7,9 +7,11 @@
 
 import * as bunyan from 'bunyan';
 import * as yargs from 'yargs';
-import { injectable } from 'inversify';
+import { inject, injectable } from 'inversify';
 import { LogLevel } from '../common/logger';
 import { ILoggerServer, ILoggerClient } from '../common/logger-protocol';
+import { LoggerServerOptions } from '../common/logger-protocol';
+
```
Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: No need for that newline here

**Comment by hexa00** [comment]
@lmcgupe  don't merge this yet 

I forgot there's something wrong you're not taking into account the loglevel passed in argument to yargs we need to fix this

File: packages/core/src/node/bunyan-logger-server.ts [hunk 13]
```diff
@@ -42,11 +43,9 @@ export class BunyanLoggerServer implements ILoggerServer {
         if (['trace', 'debug', 'info', 'warn', 'error', 'fatal'].indexOf(logLevel) < 0) {
             logLevel = 'info';
         }
```
Comment (hexa00) [inline packages/core/src/node/bunyan-logger-server.ts]: This does notthing now... we need to move that where the loglevel is set

File: packages/core/src/node/logger-backend-module.ts [hunk 14]
```diff
@@ -7,15 +7,24 @@
...
     bind(ILogger).to(Logger).inSingletonScope();
     bind(LoggerWatcher).toSelf().inSingletonScope();
     bind(ILoggerServer).to(BunyanLoggerServer).inSingletonScope();
+    bind(LoggerServerOptions).toDynamicValue(ctx => {
+        let logLevel = yargs.argv.loglevel;
+        if (['trace', 'debug', 'info', 'warn', 'error', 'fatal'].indexOf(logLevel) < 0) {
+            logLevel = 'info';
+        }
+        return JSON.parse('{ "name": "Theia",' + '"level":"' + logLevel + '" }');
+    }
```
Comment (hexa00) [inline packages/core/src/node/logger-backend-module.ts]: Did we not discuss this JSON.parse thing already?

File: packages/core/src/node/logger-backend-module.ts [hunk 15]
```diff
@@ -7,15 +7,24 @@
 
 import { ContainerModule, Container } from 'inversify';
 import { ConnectionHandler, JsonRpcConnectionHandler } from "../common/messaging";
-import { ILogger, LoggerFactory, LoggerOptions, Logger } from '../common/logger';
-import { ILoggerServer, ILoggerClient, loggerPath } from '../common/logger-protocol';
+import { ILogger, LoggerFactory, Logger, LoggerOptions } from '../common/logger';
```
Comment (hexa00) [inline packages/core/src/node/logger-backend-module.ts]: Why change the order here of Logger / LoggerOptions
